### PR TITLE
BFB porting check_values to scream

### PR DESCRIPTION
**Pull request by xyuan**
porting check_values to scream cpp

BFB

### CODE REVIEW

File: components/scream/src/physics/p3/p3_functions_check_values_impl.hpp [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+{
+  constexpr Scalar T_low  = 173.;
+  constexpr Scalar T_high = 323.;
+  constexpr Scalar Q_high = 40.e-3;
+  constexpr Scalar Q_low  = 0.;
+
+  Int kmin, kmax;
+  util::set_min_max(kstart, kend, kmin, kmax, Spack::n);
+
+
```
Comment (jgfouca) [inline components/scream/src/physics/p3/p3_functions_check_values_impl.hpp]: Remove extra newline

File: components/scream/src/physics/p3/p3_functions_check_values_impl.hpp [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  The value 'source_ind' indicates the approximate location in 'p3_main'
+  from where 'check_values' was called before it resulted in a trap.
+  -----------------------------------------------------------------------------------
+*/
+template <typename S, typename D>
+KOKKOS_FUNCTION
+void Functions<S,D>
+::check_values(const view_1d<Spack>& qv, const view_1d<Spack>& temp, 
+               const Int& kstart, const Int& kend, const MemberType& team, 
+               view_1d<Smask> qv_out_bounds, view_1d<Smask> t_out_bounds)
```
Comment (jgfouca) [inline components/scream/src/physics/p3/p3_functions_check_values_impl.hpp]: The signature of this function should be:

```
::check_values(const view_1d<const Spack>& qv, const view_1d<const Spack>& temp, 
               const Int& kstart, const Int& kend, const MemberType& team, 
               const view_1d<Smask>& qv_out_bounds, const view_1d<Smask>& t_out_bounds)
```

File: components/scream/src/physics/p3/p3_functions_check_values_impl.hpp [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+{
+  constexpr Scalar T_low  = 173.;
```
Comment (jgfouca) [inline components/scream/src/physics/p3/p3_functions_check_values_impl.hpp]: These should probably be defined in p3_constants.

File: components/scream/src/physics/p3/p3_functions_f90.cpp [hunk 4]
```diff
@@ -2271,5 +2317,87 @@ void ice_water_conservation_f(Real qitot_, Real qidep_, Real qinuc_, Real qiberg
...
+    view_1d qv_d{t_d[0]},
+            temp_d{t_d[1]};
+
+    P3F::check_values(qv_d, temp_d, kstart, kend, team, qv_out_bounds_d, t_out_bounds_d);
+
+  });
+
+  Kokkos::deep_copy(qv_out_bounds_h, qv_out_bounds_d);
+  Kokkos::deep_copy(t_out_bounds_h, t_out_bounds_d);
+
```
Comment (jgfouca) [inline components/scream/src/physics/p3/p3_functions_f90.cpp]: Everything below this line should be in the main implementation, not the bridge.

File: components/scream/src/physics/p3/tests/p3_check_values_unit_tests.cpp [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+static void run_check_values_bfb()
+{
+  const std::array< std::pair<Real, Real>, CheckValuesData::NUM_ARRAYS > ranges = {
+    std::make_pair(-4.056E-01, 1.153E+00), // qv
+    std::make_pair( 1.000E+02, 5.000E+02), // temp
+  };
+
+  CheckValuesData cvd_fortran[] = {
+    // kts_, kte_, timestepcount_, source_ind_, force_abort_, ranges
+    CheckValuesData(1,  72,   2,   100,   false,  ranges),
```
Comment (jgfouca) [inline components/scream/src/physics/p3/tests/p3_check_values_unit_tests.cpp]: There should be more variation in the inputs for better coverage. Have differnent starting/stopping k, toggled the boolean flags to different values, etc.

File: components/scream/src/physics/p3/p3_functions_check_values_impl.hpp [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    const int pk = kmin + pk_;
+
+    const auto t_gt_low_bound   = temp(pk) > T_low;
+    const auto t_lt_high_bound   = temp(pk) < T_high;
+
+    const auto qv_gt_low_bound  = qv(pk) > Q_low;
+    const auto qv_lt_high_bound = qv(pk) < Q_high;
+
+    t_out_bounds(pk)  = !(t_gt_low_bound && t_lt_high_bound);
+    qv_out_bounds(pk) = !(qv_gt_low_bound && qv_lt_high_bound);
```
Comment (jgfouca) [inline components/scream/src/physics/p3/p3_functions_check_values_impl.hpp]: Use scream_krequire to implement error checking here.

**Comment by xyuan** [comment]
close this PR, and redo it based on the latest master

**Comment by PeterCaldwell** [comment]
@xyuan - out of curiosity, why are you closing this PR and creating a new one instead of `git rebase master`?

**Comment by xyuan** [comment]
Yes, I prefer redoing this PR.

**Pull request closed without merging**